contents/codeQuality/CodeQualityMetrics.md

```diff
@@ -0,0 +1,98 @@
+
+> Since quantitative measurements are essential in all sciences, there is a continuous effort by computer science practitioners and theoreticians to bring similar approaches to software development.
+> The goal is obtaining objective, reproducible and quantifiable measurements, which may have numerous valuable applications in schedule and budget planning, cost estimation, quality assurance testing, software debugging, software performance optimization, and optimal personnel task assignments.
+>
+> https://en.wikibooks.org/wiki/Introduction_to_Software_Engineering/Quality/Metrics
+
+In fact, [the Boeing 777 project attributed its success to excellent usage of Quality Metrics (both code and non-code)](https://thesai.org/Downloads/Volume3No1/Paper%2021-Survey%20on%20Impact%20of%20Software%20Metrics%20on%20Software%20Quality.pdf).
+
+*Note that the term "Code Quality" here is scoped to be the readability, understandability, and maintainability of the code. While performance can be considered a legitimate basis for some metrics (e.g. time taken to run certain operations, based on profiling), it is not considered for this purpose.*
+
+## Example of Metrics
+
+### Complexity Metrics
+
+Complexity metrics measures how "complex" methods, classes, packages, etc. are. It is arguably the most useful metric for the largest number of developer because it is the easiest to grasp, the most directly relevant to the coding activity, and applicable to most/all programming languages.
+
+- The most well-known complexity metric is **Cyclomatic Complexity (CC)**, invented by Thomas McCabe in 1976. [Here](http://www.whiteboxtest.com/cyclomatic-complexity.php) is an excellent explanation on how to make use of it to measure your code quality. CC corresponds to the **minimum number of test cases needed to achieve 100% branch coverage**.
+- An alternative, less well-known, harder to calculate complexity metric is **NPath Complexity (NC)**, invented by Brian A. Nejmeh in 1988. NC corresponds to the **minimum number of test cases needed to achieve 100% path coverage**.
+
+> [Here](http://www.literateprogramming.com/mccabe.pdf) is the original paper by McCabe, and [here](http://delivery.acm.org/10.1145/50000/42379/p188-nejmeh.pdf?ip=137.132.250.8&id=42379&acc=ACTIVE%20SERVICE&key=FF6731C4D3E3CFFF%2EBB5EB8D2067C1662%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&CFID=726592962&CFTOKEN=31567842&__acm__=1486743928_30f8e35ed72667612c6449afe5db3998) is the original paper by Nejmeh.
+
+This complexity metrics can be extended to class level (e.g. summing or averaging the complexity values of all methods in a class), package level, or even project level.
+
+### Class Design Metrics
+
+Class design metrics measures how "well-designed" a class is. Metrics on this level are more applicable for QA team members and software architects, but still holds some relevance to students and junior developers who have to design a class-level API.
```

Comment:
In essence, we have learnt how "high cohesion, low coupling, promoting encapsulation, information hiding" are good SE principles. Some of the class design metrics can measure those (that's how powerful metrics are!) and well-designed classes conforms to those, which subsequently makes reuse, refactoring, and maintenance easier.

Complexity is a different concept as it determines just how much effort is needed to understand the class, although it has some effect to how well-designed a class is (e.g the WMC metric).

In any case, it's all explained in the chapter now.